app/src/main/java/com/zulip/android/activities/ZulipActivity.java

```diff
@@ -535,6 +538,146 @@ public Cursor runQuery(CharSequence charSequence) {
+
+        if (Intent.ACTION_SEND.equals(action) && type != null) {
+            if (type.startsWith("image/")) {
+                handleSentImage(intent); // Handle single image being sent
+                Log.e("intent", action);
+            }
+        }
+    }
+
+    /**
+     * Function invoked when a user shares an image with the zulip app
+     * @param intent passed to the activity with action SEND
+     */
+    private void handleSentImage(Intent intent) {
+        Uri imageUri = (Uri) intent.getParcelableExtra(Intent.EXTRA_STREAM);
+        if (imageUri != null) {
+            // Update UI to reflect image is being loaded
+            // hide fab and display chatbox
+            displayFAB(false);
+            displayChatBox(true);
+            String loadingMsg = getResources().getString(R.string.uploading_message);
+            sendingMessage(true, loadingMsg);
+
+            // get actual file path
+            String imageFilePath = FilePathHelper.getPath(this, imageUri);
+
+            // upload the file asynchronously to the server
+            uploadFile(imageFilePath);
+        } else {
+            Toast.makeText(this, "Could not find the image", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    /**
+     * Function to upload file asynchronously to the server using retrofit callback
+     * upload {@link com.zulip.android.service.ZulipServices#upload(RequestBody, MultipartBody.Part)}
+     * @param filePath on local storage
+     */
+    private void uploadFile(String filePath) {
+        File file = new File(filePath);
+
+        // create RequestBody instance from file
+        RequestBody requestFile =
+                RequestBody.create(MediaType.parse("multipart/form-data"), file);
+
+        // MultipartBody.Part is used to send also the actual file name
+        MultipartBody.Part body =
+                MultipartBody.Part.createFormData("picture", file.getName(), requestFile);
+
+        // add another part within the multipart request
+        String descriptionString = "hello, this is description speaking";
+        RequestBody description =
+                RequestBody.create(
+                        MediaType.parse("multipart/form-data"), descriptionString);
+
+        final String loadingMsg = getResources().getString(R.string.uploading_message);
+
+        // finally, execute the request
+        // create upload service client
+        Call<ResponseBody> call = ((ZulipApp) getApplicationContext()).getZulipServices().upload(description, body);
+        call.enqueue(new Callback<ResponseBody>() {
+            @Override
+            public void onResponse(Call<ResponseBody> call,
+                                   Response<ResponseBody> response) {
+                if (response.message().equals("OK")) {
+                    String filePathOnServer = "";
+                    try {
+                        String jsonResponse = getJsonResponseForUpload(response.body().charStream());
```

Comment:
Thank you very much for the rigorous code view, I didn't know about gson and the best practices followed when consuming RESTful apis using retrofit. So I when through a couple of tutorials and made a 'UploadResponse` pojo and used it to deserialize the response.